src/core/localize.rkt

```diff
@@ -103,36 +104,35 @@
 
 ; Compute local error or each sampled point at each node in `prog`.
 (define (compute-local-errors subexprss ctx)
-  (define spec-list
-    (for*/list ([subexprs (in-list subexprss)] [subexpr (in-list subexprs)])
-      (prog->spec subexpr)))
+  (define exprs-list (append* subexprss)) ; unroll subexprss
   (define ctx-list
-    (for*/list ([subexprs (in-list subexprss)] [subexpr (in-list subexprs)])
+    (for/list ([subexpr (in-list exprs-list)])
       (struct-copy context ctx [repr (repr-of subexpr ctx)])))
 
-  (define subexprs-fn (eval-progs-real spec-list ctx-list))
+  (define expr-batch (progs->batch exprs-list #:vars (context-vars (first ctx-list))))
+  (define nodes (batch-nodes expr-batch))
+  (define roots (batch-roots expr-batch))
+
+  (define subexprs-fn (eval-progs-real (map prog->spec exprs-list) ctx-list))
 
   ; Mutable error hack, this is bad
-  (define errs
-    (make-hash (for*/list ([subexprs (in-list subexprss)] [subexpr (in-list subexprs)])
-                 (cons subexpr '()))))
+  (define errs (make-hash (map list exprs-list)))
 
   (for ([(pt ex) (in-pcontext (*pcontext*))])
-    (define exacts (apply subexprs-fn pt))
-    (define exacts-hash (make-immutable-hash (map cons (apply append subexprss) exacts)))
-    (for* ([subexprs (in-list subexprss)] [expr (in-list subexprs)])
+    (define exacts (list->vector (apply subexprs-fn pt)))
+    (for ([expr (in-list exprs-list)] [root (in-vector roots)] [exact (in-vector exacts)])
```

Comment:
I think this is fine for now but in the longer term probably it would be faster to accumulate the results in a vector instead of in a hash table from expr to errors.